pkg/distributor/distributor.go

```diff
@@ -241,18 +253,41 @@ func (d *Distributor) Push(ctx context.Context, req *client.WriteRequest) (*clie
 		return nil, err
 	}
 
+	d.haFlag = true
 	var lastPartialErr error
 
+	// Check the HA cluster and instance labels.
+	if d.haFlag && len(req.Timeseries) > 0 {
+		ts := req.Timeseries[0]
+		cluster, instance, err := removeHALabels(&ts.Labels)
+		if err != nil && (cluster != "" && instance != "") {
+			return nil, err
+		}
+		req.Timeseries[0].Labels = ts.Labels
+
+		// Lookup the cluster/instance here to see if we want to accept this sample
+		if inst := d.instances.LookupInstance(cluster, instance); inst.instance != instance && (cluster != "" && instance != "") { // Don't accept the sample.
+			// If we assume all samples will have the same cluster/instance labels we can return an error here.
+			// Should we just silently drop here?
+			return nil, fmt.Errorf("dropping sample")
```

Comment:
I think we do want to silently drop, but we probably want to check acceptance (or not) after the validation loop below.  And the you can check all the timeseries have the same cluster/instance label in that loop too, for extra points.